targets/psa_switch/externs/psa_internet_checksum.cpp

```diff
@@ -0,0 +1,133 @@
+/* Copyright 2021 SYRMIA LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/*
+ * Dusan Krdzic (dusan.krdzic@syrmia.com)
+ *
+ */
+
+
+#include "psa_internet_checksum.h"
+#include <bm/bm_sim/_assert.h>
+
+
+namespace {
+
+void
+concatenate_fields(const std::vector<bm::Field> &fields, bm::Data &input) {
+    uint16_t n_bits = 0;
+    bm::Data field_shl;
+
+    // Concatenate fields in one single data
+    for (int i = fields.size() - 1; i >= 0; i--) {
+        field_shl.shift_left(fields.at(i), n_bits);
+        input.add(input, field_shl);
+        n_bits += fields.at(i).get_nbits();
+    }
+
+    _BM_ASSERT(n_bits % 16 == 0);
+}
+
+uint16_t
+ones_complement_sum(uint16_t x, uint16_t y) {
+    uint32_t ret = (uint32_t) x + (uint32_t) y;
+    if (ret >= 0x10000)
+        ret++;
+
+    return ret;
+}
+
+}  // namespace
+
+namespace bm {
+
+namespace psa {
+
+void
+PSA_InternetChecksum::init() {
+    clear();
+}
+
+void
+PSA_InternetChecksum::get(Field &dst) const {
+    dst.set(static_cast<uint16_t>(~sum));
+}
+
+void
+PSA_InternetChecksum::get_state(Field &dst) const {
+    dst.set(sum);
+}
+
+void
+PSA_InternetChecksum::get_verify(Field &dst, Field &equOp) const {
+    dst.set(equOp.get<uint16_t>() == static_cast<uint16_t>(~sum));
+}
+
+void
+PSA_InternetChecksum::set_state(const Data &src) {
+    sum = src.get<uint16_t>();
+}
+
+void
+PSA_InternetChecksum::clear() {
+    sum = 0;
+}
+
+void
+PSA_InternetChecksum::add(const std::vector<Field> &fields) {
+    bm::Data input(0);
+    bm::Data chunk;
+
+    concatenate_fields(fields, input);
+
+    while (!input.test_eq(0)) {
+        chunk.bit_and(input, 0xFFFF);
```

Comment:
I'd rather have something like this defined:
```
static const mask Data(0xffff);
...
chunk.bit_and(input, mask)
```

I don't feel like there is a need for the new `bit_and` variant.

same comment below